### PR TITLE
Matmul 2D Tiling

### PR DESCRIPTION
**Pull request by louisfd**
Matmul 2D Tiling

### CODE REVIEW

**Review by nathanielsimard — CHANGES_REQUESTED** [review]

File: burn-wgpu/src/kernel/matmul.rs [hunk 1]
```diff
@@ -1,16 +1,49 @@
...
+use super::{build_info, DynamicKernelSettings, SourceTemplate, StaticKernel};
 use crate::{context::WorkGroup, element::WgpuElement, kernel_wgsl, tensor::WgpuTensor};
 use burn_tensor::Shape;
 
-const BLOCK_SIZE: usize = 16;
+const B_M: usize = 128;
+const B_N: usize = 128;
+const B_K: usize = 8;
+const T_M: usize = 8;
+const T_N: usize = 8;
```
Comment (antimora) [inline burn-wgpu/src/kernel/matmul.rs:11]: Could we add comments for each constant to signify what B_M stands for, e.g.? The names to short to know what they're.

File: burn-wgpu/src/kernel/matmul.rs [hunk 2]
```diff
@@ -1,16 +1,49 @@
...
 ) -> WgpuTensor<E, D> {
+    matmul_tiling_2d(lhs, rhs)
+}
+
+pub fn matmul_tiling_2d<E: WgpuElement, const D: usize>(
+    lhs: WgpuTensor<E, D>,
+    rhs: WgpuTensor<E, D>,
+) -> WgpuTensor<E, D> {
+    assert!(B_K <= min(B_M, B_N), "B_K must be smaller than both B_M and B_M, otherwise there won't be enough threads to fill shared memory. ");
+    assert!(B_K * max(B_M, B_N) <= 8192, "B_K x B_M and B_K x B_N must be smaller or equal than 8192, otherwise shared memory limit will be busted. ");
```
Comment (antimora) [inline burn-wgpu/src/kernel/matmul.rs:44]: Can we extract the magic number 8192 into a const with a name and comment?

File: burn-wgpu/src/template/matmul_blocktiling_2d.wgsl [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    let actual_T_M = min(B_M - thread_row, T_M);
+    let actual_T_N = min(B_N - thread_col, T_N);
+
+    var results: array<{{ elem }}, T_M_X_T_N>;
+    var register_M: array<{{ elem }}, T_M>;
+    var register_N: array<{{ elem }}, T_N>;
+
+    for (var k = 0u; k < K; k += B_K) { 
+        let sm_limit = min(B_K, K - k);
+
```
Comment (antimora) [inline burn-wgpu/src/template/matmul_blocktiling_2d.wgsl:81]: For big blocks of code, can we add a brief intention comment? It's a good practice in general.

File: burn-wgpu/src/template/matmul_blocktiling_2d.wgsl [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                if current_row < sm_limit {
+                    let rhs_sm_position = current_row * B_N + current_col; 
+                    let rhs_position = offset_rhs + (k + current_row) * n_cols + current_col;
+                    shared_rhs[rhs_sm_position] = rhs[rhs_position];
+                }
+            }
+        }
+
+        workgroupBarrier();
+
```
Comment (antimora) [inline burn-wgpu/src/template/matmul_blocktiling_2d.wgsl:102]: Same

File: burn-wgpu/src/template/matmul_blocktiling_2d.wgsl [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            for (var res_idx_M = 0u; res_idx_M < actual_T_M; res_idx_M++) {
+                for (var res_idx_N = 0u; res_idx_N < actual_T_N; res_idx_N++) {
+                    results[res_idx_M * actual_T_N + res_idx_N] += register_M[res_idx_M] * register_N[res_idx_N];
+                }
+            }
+        }
+        
+        workgroupBarrier();
+    }
+
```
Comment (antimora) [inline burn-wgpu/src/template/matmul_blocktiling_2d.wgsl:121]: Same

**Review by antimora — CHANGES_REQUESTED** [review]
Overall looks good but lacks comments. I would encourage to add intention comments above any big blocks to guide the reader.

File: burn-wgpu/src/kernel/matmul.rs [hunk 6]
```diff
@@ -1,16 +1,65 @@
...
+const WORKGROUP_SIZE_X: usize = 16;
+// WORKGROUP_SIZE_Y = ceil(B_N / T_N)
+const WORKGROUP_SIZE_Y: usize = 16;
+
+const MAX_SHARED_MEMORY_SIZE: usize = 8192;
+
+kernel_wgsl!(MatmulTiling2DRaw, "../template/matmul_blocktiling_2d.wgsl");
+
+struct MatmulTiling2D;
+// struct MatmulCaching;
```
Comment (antimora) [inline burn-wgpu/src/kernel/matmul.rs:32]: Looks like this dead code is left out:  `// struct MatmulCaching;`

**Review by antimora — APPROVED** [review]
Looks much better with comments thanks for adding. 

There is one commented code you'd want to remove it if not needed.

**Review by nathanielsimard — APPROVED** [review]
🚀🔥🎉

**Pull request merged**